library/core/src/num/nonzero.rs

```diff
@@ -1,200 +1,472 @@
+{
+    #[inline]
+    fn clone(&self) -> Self {
+        // SAFETY: A `NonZero` is guaranteed to only contain non-zero values.
+        unsafe { Self(self.0) }
+    }
+}
 
-            #[stable(feature = "nonzero_bitor", since = "1.45.0")]
-            #[rustc_const_unstable(feature = "const_ops", issue = "90080")]
-            impl const BitOr<$Int> for $Ty {
-                type Output = Self;
-                #[inline]
-                fn bitor(self, rhs: $Int) -> Self::Output {
-                    // SAFETY: since `self` is nonzero, the result of the
-                    // bitwise-or will be nonzero regardless of the value of
-                    // `rhs`.
-                    unsafe { $Ty::new_unchecked(self.get() | rhs) }
-                }
-            }
+#[unstable(feature = "generic_nonzero", issue = "82363")]
+impl<T: ZeroablePrimitive> const PartialEq for NonZero<T>
+where
+    T: ~const PartialEq,
+{
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        // SAFETY: A `NonZero` is guaranteed to only contain non-zero values.
+        unsafe { self.0 == other.0 }
+    }
 
-            #[stable(feature = "nonzero_bitor", since = "1.45.0")]
-            #[rustc_const_unstable(feature = "const_ops", issue = "90080")]
-            impl const BitOr<$Ty> for $Int {
-                type Output = $Ty;
-                #[inline]
-                fn bitor(self, rhs: $Ty) -> Self::Output {
-                    // SAFETY: since `rhs` is nonzero, the result of the
-                    // bitwise-or will be nonzero regardless of the value of
-                    // `self`.
-                    unsafe { $Ty::new_unchecked(self | rhs.get()) }
-                }
-            }
+    #[inline]
+    fn ne(&self, other: &Self) -> bool {
+        // SAFETY: A `NonZero` is guaranteed to only contain non-zero values.
+        unsafe { self.0 != other.0 }
+    }
+}
 
-            #[stable(feature = "nonzero_bitor", since = "1.45.0")]
-            #[rustc_const_unstable(feature = "const_ops", issue = "90080")]
-            impl const BitOrAssign for $Ty {
-                #[inline]
-                fn bitor_assign(&mut self, rhs: Self) {
-                    *self = *self | rhs;
-                }
-            }
+#[stable(feature = "nonzero", since = "1.28.0")]
+#[rustc_const_unstable(feature = "const_ops", issue = "90080")]
+impl<T: ZeroablePrimitive> const StructuralPartialEq for NonZero<T> {}
 
-            #[stable(feature = "nonzero_bitor", since = "1.45.0")]
-            #[rustc_const_unstable(feature = "const_ops", issue = "90080")]
-            impl const BitOrAssign<$Int> for $Ty {
+#[unstable(feature = "generic_nonzero", issue = "82363")]
+impl<T: ZeroablePrimitive> const PartialOrd for NonZero<T>
+where
+    T: ~const PartialOrd,
+{
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        // SAFETY: A `NonZero` is guaranteed to only contain non-zero values.
+        unsafe { self.0.partial_cmp(&other.0) }
```

Comment:
This and the rest of this impl doesn't seem like it needs any unsafe blocks.

Reply:
These do need `unsafe`, otherwise I get this error:

```
borrow of layout constrained field with interior mutability is unsafe and requires unsafe block
```

Reply:
Which is also why these traits cannot be derived and has to be implemented manually.